### PR TITLE
r/virtual_machine: Refactor tests

### PR DESCRIPTION
**Pull request by vancluever**
Maintaining this file has become increasingly difficult. As such, 
I have re-factored the tests in the style as we are doing most
of the new tests - highly functional, static configs with a tolerable
level of repetition, and all tests set up as subtests, so every test is
easy to locate as part of the larger test case list.

I have also removed a few tests that currently do not function, for one
reason or another. A couple of these are due to present bugs in the
resource, but as pending work may deprecate the resource in its current
form, I think maintenance effort is better spent elsewhere as they are
not currently reported business critical bugs that are rendering the
resource unusable.

The tests were:

* A test with a CDROM virtual device - this test was panicking. As it's
hard to imagine who is using VMs in TF with CDROM devices attached, I
think this is a bit of an edge case.
* A test for a NIC with DHCP. In theory this should work, but it's hard
to test properly without dealing with races in VMware tools, so the test
itself probably did not have much value.
* A test to detach disks added out-of-band. This is indeed a bug as
`Read` does not seem to be picking up the extra disk and triggering the
needed diff. Again, this is something that will be picked up on the
refactor, or might look completely different afterwards.

Some other tests were consolidated - memory and CPU changes are now one
test, and the individual `keep_on_remove` test for disk devices was
dropped as we now test this with the test that adds an external disk via
`vsphere_virtual_disk`, where a lack of `keep_on_remove` on the attachment
actually causes issues during destroy as the disk is removed
prematurely, with the virtual machine.

### CODE REVIEW

File: vsphere/resource_vsphere_virtual_machine_test.go [hunk 1]
```diff
@@ -1,1726 +1,1548 @@
...
+		}
+		return nil
 	}
+}
 
-	if v := os.Getenv("VSPHERE_IPV4_ADDRESS"); v == "" {
-		t.Fatal("env variable VSPHERE_IPV4_ADDRESS must be set for acceptance tests")
+// testAccResourceVSphereVirtualMachineCheckHostname is a check to check for a
+// VirtualMachine's hostname. The check uses guest info, so VMware tools needs
+// to be installed.
```
Comment (mbfrahry) [inline vsphere/resource_vsphere_virtual_machine_test.go]: What would happen if VMware tools weren't installed? Is there a way to check if the tools are installed before running tests that call this?

File: vsphere/resource_vsphere_virtual_machine_test.go [hunk 2]
```diff
@@ -1,1726 +1,1548 @@
...
+// backings, and expects them in the exact order outlined in the function.
+func testAccResourceVSphereVirtualMachineCheckExtraDisks() resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		props, err := testGetVirtualMachineProperties(s, "vm")
+		if err != nil {
+			return err
+		}
+
+		var foundEager, foundLazy, foundThin bool
+		expectedEagerName := "terraform-test-extra-eager.vmdk"
```
Comment (mbfrahry) [inline vsphere/resource_vsphere_virtual_machine_test.go]: Is there any chance for these names to change? I'm thinking we make these constants

File: vsphere/resource_vsphere_virtual_machine_test.go [hunk 3]
```diff
@@ -1,1726 +1,1548 @@
...
+		expectedLazyName := "terraform-test-extra-lazy.vmdk"
```
Comment (mbfrahry) [inline vsphere/resource_vsphere_virtual_machine_test.go]: Thoughts on making these constants?

File: vsphere/resource_vsphere_virtual_machine_test.go [hunk 4]
```diff
@@ -1,1726 +1,1548 @@
...
+		expectedThinName := "terraform-test-extra-thin.vmdk"
```
Comment (mbfrahry) [inline vsphere/resource_vsphere_virtual_machine_test.go]: Thoughts on making this a constant?

File: vsphere/resource_vsphere_virtual_machine_test.go [hunk 5]
```diff
@@ -1,1726 +1,1548 @@
...
+// for the MAC address defined in the static MAC address test on the first
+// network interface.
+func testAccResourceVSphereVirtualMachineCheckStaticMACAddr() resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		props, err := testGetVirtualMachineProperties(s, "vm")
+		if err != nil {
+			return err
+		}
+		actual := props.Guest.Net[0].MacAddress
+		expected := "06:5c:89:2b:a0:64"
```
Comment (mbfrahry) [inline vsphere/resource_vsphere_virtual_machine_test.go]: Thoughts on making this a constant?

File: vsphere/resource_vsphere_virtual_machine_test.go [hunk 6]
```diff
@@ -1,1726 +1,1548 @@
...
+
+// testAccResourceVSphereVirtualMachineCheckAnnotation is a check to ensure
+// that a VM's annotation is correctly set in the annotation test.
+func testAccResourceVSphereVirtualMachineCheckAnnotation() resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		props, err := testGetVirtualMachineProperties(s, "vm")
+		if err != nil {
+			return err
+		}
+		expected := "Managed by Terraform"
```
Comment (mbfrahry) [inline vsphere/resource_vsphere_virtual_machine_test.go]: Thoughts on constant?

File: vsphere/resource_vsphere_virtual_machine_test.go [hunk 7]
```diff
@@ -1,1726 +1,1548 @@
...
+		actual := props.Config.Annotation
```
Comment (mbfrahry) [inline vsphere/resource_vsphere_virtual_machine_test.go]: And constant here?

File: vsphere/resource_vsphere_virtual_machine_test.go (hunk 1 above)
Comment (vancluever) [reply vsphere/resource_vsphere_virtual_machine_test.go]: If VMware tools is not installed, I think this test will either fail due to no data being present or panic, depending on how well populated `Guest` is populated in this scenario. In any case, it will fail, which is what we want it to do.

A good chunk of the resource right now really depends on VMware tools being installed on the template that you are basing it off of, and all of our tests (and most of the old tests as well) depend on a template. There might be a scenario in the future where we have a good use case where we need to deploy something with Terraform without VMware tools, like a virtual appliance - when we have that, we will probably need to create an environment variable that can be used to skip tests like these... but for now we don't have a test that would consume it, so it's not really necessary.

File: vsphere/resource_vsphere_virtual_machine_test.go (hunk 7 above)
Comment (vancluever) [reply vsphere/resource_vsphere_virtual_machine_test.go]: This is the actual value which we need to pull from the VM state (`Config` here denotes the configuration section of the retrieved properties).

Comment (mbfrahry) [reply vsphere/resource_vsphere_virtual_machine_test.go]: My bad. Didn't notice the lack of ""

**Comment by vancluever** [comment]
Thanks @mbfrahry! Just needed to review a couple of breaking tests before I merged here :+1:
